File: c/grains/grains.c
#include "grains.h"

/*
 * See the following wikipedia article to understand how the following two
 * 	formulas are derived:
 *
 * 	https://en.wikipedia.org/wiki/Wheat_and_chessboard_problem#Solutions
 */

uint64_t square(uint8_t index)
{
	return pow(2, index-1);
}

uint64_t total(void)
{
	return powl(2, 64) - 1;
}

File: go/grains/grains.go
package grains

import (
    "errors"
    "math"
)

func Square(number int) (uint64, error) {
    if number < 1 || number > 64 {
        return 0, errors.New("number must be between 1 and 64 inclusively")
    }

	return uint64(math.Pow(2, float64(number - 1))), nil
}

func Total() uint64 {
    var total uint64
    for i := 1; i <= 64; i++ {
        n, _ := Square(i)
        total += n
    }

	return total
}
